csharp/src/ProtocolBuffers/JsonFormatter.cs

```diff
@@ -0,0 +1,578 @@
+            "\\u0080", "\\u0081", "\\u0082", "\\u0083",  // 0x80
+          "\\u0084", "\\u0085", "\\u0086", "\\u0087",
+          "\\u0088", "\\u0089", "\\u008a", "\\u008b",
+          "\\u008c", "\\u008d", "\\u008e", "\\u008f",
+          "\\u0090", "\\u0091", "\\u0092", "\\u0093",  // 0x90
+          "\\u0094", "\\u0095", "\\u0096", "\\u0097",
+          "\\u0098", "\\u0099", "\\u009a", "\\u009b",
+          "\\u009c", "\\u009d", "\\u009e", "\\u009f"
+        };
+
+        static JsonFormatter()
+        {
+            for (int i = 0; i < CommonRepresentations.Length; i++)
+            {
+                if (CommonRepresentations[i] == "")
+                {
+                    CommonRepresentations[i] = ((char) i).ToString();
+                }
+            }
+        }
+
+        private readonly Settings settings;
+
+        public JsonFormatter(Settings settings)
+        {
+            this.settings = settings;
+        }
+
+        public string Format(IReflectedMessage message)
+        {
+            ThrowHelper.ThrowIfNull(message, "message");
+            StringBuilder builder = new StringBuilder();
+            WriteMessage(builder, message);
+            return builder.ToString();
+        }
+
+        private void WriteMessage(StringBuilder builder, IReflectedMessage message)
+        {
+            if (message == null)
+            {
+                WriteNull(builder);
+                return;
+            }
+            builder.Append("{ ");
+            var fields = message.Fields;
+            bool first = true;
+            // First non-oneof fields
+            foreach (var accessor in fields.Accessors)
+            {
+                var descriptor = accessor.Descriptor;
+                // Oneofs are written later
+                if (descriptor.ContainingOneof != null)
+                {
+                    continue;
+                }
+                // Omit default values unless we're asked to format them
+                object value = accessor.GetValue(message);
+                if (!settings.FormatDefaultValues && IsDefaultValue(accessor, value))
+                {
+                    continue;
+                }
+                // Omit awkward (single) values such as unknown enum values
+                if (!descriptor.IsRepeated && !descriptor.IsMap && !CanWriteSingleValue(accessor.Descriptor, value))
+                {
+                    continue;
+                }
+
+                // Okay, all tests complete: let's write the field value...
+                if (!first)
+                {
+                    builder.Append(", ");
+                }
+                WriteString(builder, ToCamelCase(accessor.Descriptor.Name));
+                builder.Append(": ");
+                WriteValue(builder, accessor, value);
+                first = false;
+            }
+
+            // Now oneofs
+            foreach (var accessor in fields.Oneofs)
```

Comment:
Well, that code would be *considerably* uglier. I can try it if you want, but bearing in mind that the field numbers aren't even visible in the JSON, and oneofs already behave a bit weirdly in JSON, I think I'd prefer to keep it like this. I would expect different implementations to make different decisions on this, when it comes to ordering. (I'm sure I saw something somewhere saying that the ordering didn't matter for JSON, but I can't find it now.)